### ISSUE TITLE
Multi-label training

### ISSUE THREAD

**Issue by zachluo** [ultralytics/yolov5#5560]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar feature requests.

### Description

Hi, 
I really appreciate your great work for the cv community. Recently I realized the multi-label training and found that it worked well on my customized dataset. Specifically, I squeezed the labels before building the targets. The implementations are listed as follows. Any discussion is welcome.

    from collections import OrderedDict

    def squeeze_targets(self, targets):
        # input targets(image,class,x,y,w,h)

        # copy the input since we need to modify it
        targets_copy = targets.clone()
        # convert the key (image, x, y, w, h) to string for hashing
        keys = [np.array2string(key) for key in targets[:, [0, 2, 3, 4]].cpu().numpy()]

        # store all the classes for each unique box
        classes = OrderedDict()
        classes[keys[0]] = [targets[0, 1].long()]

        new_targets = [targets_copy[0]]
        for i, t in enumerate(targets[1:]):
            k = keys[i + 1]
            if k in classes.keys():
                classes[k].append(t[1].long())
            else:
                classes[k] = [t[1].long()]
                # here we change the class id to the row index of class array
                targets_copy[i + 1, 1] = len(new_targets)
                new_targets.append(targets_copy[i + 1])

        new_targets = torch.stack(new_targets)
        return new_targets, list(classes.values())

    def __call__(self, p, targets):  # predictions, targets, model
        device = targets.device
        lcls, lbox, lobj = torch.zeros(1, device=device), torch.zeros(1, device=device), torch.zeros(1, device=device)
        new_targets, classes = self.squeeze_targets(targets)
        tcls, tbox, indices, anchors = self.build_targets(p, new_targets)  # targets

        # Losses
        for i, pi in enumerate(p):  # layer index, layer predictions
            b, a, gj, gi = indices[i]  # image, anchor, gridy, gridx
            tobj = torch.zeros_like(pi[..., 0], device=device)  # target obj

            n = b.shape[0]  # number of targets
            if n:
                ps = pi[b, a, gj, gi]  # prediction subset corresponding to targets

                # Regression
                pxy = ps[:, :2].sigmoid() * 2. - 0.5
                pwh = (ps[:, 2:4].sigmoid() * 2) ** 2 * anchors[i]
                pbox = torch.cat((pxy, pwh), 1)  # predicted box
                iou = bbox_iou(pbox.T, tbox[i], x1y1x2y2=False, CIoU=True)  # iou(prediction, target)
                lbox += (1.0 - iou).mean()  # iou loss

                # Objectness
                tobj[b, a, gj, gi] = (1.0 - self.gr) + self.gr * iou.detach().clamp(0).type(tobj.dtype)  # iou ratio

                # Classification
                if self.nc > 1:  # cls loss (only if multiple classes)
                    t = torch.full_like(ps[:, 5:], self.cn, device=device)  # targets
                    for j in range(n):
                        for c in classes[tcls[i][j]]:
                            t[j, c] = self.cp
                    lcls += self.BCEcls(ps[:, 5:], t)  # BCE

### Use case

The training labels go to something like that.
cls_1, x1, y1, w1, h1
cls_2, x1, y1, w1, h1
cls_3, x2, y2, w2, h2
cls_4, x2, y2, w2, h2

After squeezing, the labels go to
row_num1, x1, y1, w1, h1
row_num2, x2, y2, w2, h2

[row_num1] cls_1, cls_2
[row_num2] cls_3, cls_4

### Additional

Before modification, the MAP goes to something like 95% but the prediction scores are quite low because of labels conflicts. And then it goes to 97% and the prediction scores go back normal since we allow the targets to have multiple self.cp in the same row.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@zachluo oh this is really interesting! We've had a few users ask about the best way to implement multi-label and up until now the loss function has been a problem as the labels are 'one-hot' encoded, which means all the rest of the labels would have a target value of zero. 

Can you submit a PR with this update to help everyone else working on multi-label? Please see our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/) to get started.

**Comment by zachluo** [comment]
> @zachluo oh this is really interesting! We've had a few users ask about the best way to implement multi-label and up until now the loss function has been a problem as the labels are 'one-hot' encoded, which means all the rest of the labels would have a target value of zero.
> 
> Can you submit a PR with this update to help everyone else working on multi-label? Please see our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/) to get started.

Will do in a few days:)

**Comment by glenn-jocher** [comment]
@zachluo great, thank you!

**Comment by zachluo** [comment]
> @zachluo great, thank you!

I have made a new pull request.

**Comment by glenn-jocher** [comment]
@zachluo thanks, I will take a look today!
